File: ChatGPT/Patrón de Diseño Singleton/golang/singleton.go
package main
import (
    "fmt"
    "sync"
)

type singleton struct{}
var instance *singleton
var once sync.Once

func GetInstance() *singleton {
    once.Do(func() {
        instance = &singleton{}
    })
    return instance
}

func main() {
    s1 := GetInstance()
    s2 := GetInstance()
    fmt.Printf("Puntero s1: %p\n", s1)
    fmt.Printf("Puntero s2: %p\n", s2)
    fmt.Println("¿Misma instancia?", s1 == s2)
}

File: ChatGPT/Patrón de Diseño Singleton/java/src/Singleton.java
public class Singleton {
    private static Singleton instance;
    private Singleton() {}
    public static synchronized Singleton getInstance() {
        if (instance == null) {
            instance = new Singleton();
        }
        return instance;
    }
    public static void main(String[] args) {
        Singleton s1 = Singleton.getInstance();
        Singleton s2 = Singleton.getInstance();
        System.out.println("HashCode instancia s1: " + s1.hashCode());
        System.out.println("HashCode instancia s2: " + s2.hashCode());
        System.out.println("¿Misma instancia? " + (s1 == s2));
    }
}
